plugins/upower.c

```diff
@@ -0,0 +1,471 @@
+	const char *property = "Percentage";
+
+	if (cond == OFONO_ATOM_WATCH_CONDITION_UNREGISTERED) {
+		DBG("UNREGISTERED");
+
+		modems = g_list_remove(modems, modem);
+		if (modems != NULL)
+			return;
+
+		if (upower_battery_watch) {
+			g_dbus_remove_watch(connection, upower_battery_watch);
+			upower_battery_watch = 0;
+		}
+
+		return;
+	}
+
+	DBG("REGISTERED");
+
+	/* TODO: handle removable batteries */
+
+	modems = g_list_append(modems, modem);
+
+	if (modems->next != NULL)
+		return;
+
+	upower_battery_watch = g_dbus_add_signal_watch(connection,
+						UPOWER_SERVICE,
+						battery_device_path,
+						DBUS_INTERFACE_PROPERTIES,
+						"PropertiesChanged",
+						battery_props_changed,
+						NULL, NULL);
+
+	/* Query current battery value */
+	msg = dbus_message_new_method_call(UPOWER_SERVICE, battery_device_path,
+						DBUS_PROPERTIES_INTERFACE,
+						"Get");
+	if (msg == NULL) {
+		ofono_error("%s: dbus_message_new_method failed", __func__);
+		return;
+	}
+
+	dbus_message_iter_init_append(msg, &iter);
+	dbus_message_iter_append_basic(&iter, DBUS_TYPE_STRING, &iface);
+	dbus_message_iter_append_basic(&iter, DBUS_TYPE_STRING, &property);
+
+	if (!dbus_connection_send_with_reply(connection, msg, &call, -1)) {
+		ofono_error("%s: Sending EnumerateDevices failed", __func__);
+		goto done;
+	}
+
+	dbus_pending_call_set_notify(call, get_property_reply, NULL, NULL);
+done:
+	dbus_message_unref(msg);
+}
+
+static void modemwatch(struct ofono_modem *modem, gboolean added, void *user)
+{
+
+	const char *path = ofono_modem_get_path(modem);
+
+	DBG("modem: %s, added: %d", path, added);
+
+	if (added) {
+		guint watch_id;
+
+		watch_id = __ofono_modem_add_atom_watch(modem,
+					OFONO_ATOM_TYPE_EMULATOR_HFP,
+					emulator_hfp_watch, modem, NULL);
+
+		g_hash_table_insert(modem_hfp_watches, g_strdup(path),
+						GINT_TO_POINTER(watch_id));
+
+	} else {
+		guint *watch_id = g_hash_table_lookup(modem_hfp_watches, path);
+
+		if (watch_id != NULL) {
+			__ofono_modem_remove_atom_watch(modem, *watch_id);
+			g_hash_table_remove(modem_hfp_watches, path);
+		}
+	}
+}
+
+static void call_modemwatch(struct ofono_modem *modem, void *user)
+{
+	modemwatch(modem, TRUE, user);
+}
+
+static gboolean parse_devices_reply(DBusMessage *reply)
+{
+	DBusMessageIter array, iter;
+	const char *path;
+
+	DBG("");
+
+	if (dbus_message_get_type(reply) == DBUS_MESSAGE_TYPE_ERROR) {
+		ofono_error("%s: ERROR reply to EnumerateDevices", __func__);
+		return FALSE;
+	}
+
+	if (dbus_message_iter_init(reply, &array) == FALSE) {
+		ofono_error("%s: error initializing array iter", __func__);
+		return FALSE;
+	}
+
+	if (dbus_message_iter_get_arg_type(&array) != DBUS_TYPE_ARRAY) {
+		ofono_error("%s: type != ARRAY!", __func__);
+		return FALSE;
+	}
+
+	dbus_message_iter_recurse(&array, &iter);
+
+	while (dbus_message_iter_get_arg_type(&iter) == DBUS_TYPE_OBJECT_PATH) {
+
+		dbus_message_iter_get_basic(&iter, &path);
+
+		if (strlen(path) && g_strrstr(path, "/battery_")) {
```

Comment:
Got an error here when compiling:

plugins/upower.c:353:3: error: implicit declaration of function 'strlen' [-Werror=implicit-function-declaration]
   if (strlen(path) && g_strrstr(path, "/battery_")) {
   ^

Besides, is calling strlen() really needed? I see that is not mentioned in the documentation for this function, but most g_str* functions check for null pointers.